dds/DCPS/XTypes/DynamicType.cpp

```diff
@@ -0,0 +1,128 @@
+#include "DCPS/DdsDcps_pch.h"
+#include "DynamicType.h"
+#include "MemberDescriptor.h"
+#include "TypeDescriptor.h"
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+namespace OpenDDS {
+namespace XTypes {
+
+DynamicType::DynamicType()
+  : descriptor_(new TypeDescriptor())
+{}
+
+DynamicType::~DynamicType()
+{
+  delete descriptor_;
+}
+
+void DynamicType::get_descriptor(TypeDescriptor& descriptor) const
+{
+  descriptor = *descriptor_;
+}
+
+DCPS::String DynamicType::get_name() const
+{
+  return this->descriptor_->name;
+}
+
+TypeKind DynamicType::get_kind() const
+{
+  return this->descriptor_->kind;
+}
+
+DDS::ReturnCode_t DynamicType::get_member_by_name(DynamicTypeMember_rch& member, const DCPS::String& name)
+{
+  //We have a map of members in member_by_name which takes a string and gives 0 or 1 type DTMs
+  //this should use the string as the key and put the returned DTM into member.
+  DynamicTypeMembersByName::const_iterator pos = member_by_name.find(name);
+  if (pos != member_by_name.end()) {
+    member = pos->second;
+    return DDS::RETCODE_OK;
+  } else {
+    return DDS::RETCODE_ERROR;
+  }
+}
+
+void DynamicType::get_all_members_by_name(DynamicTypeMembersByName& member)
+{
+  member = member_by_name;
+}
+
+DDS::ReturnCode_t DynamicType::get_member(DynamicTypeMember_rch& member, MemberId id)
+{
+  DynamicTypeMembersById::const_iterator pos = member_by_id.find(id);
+  if (pos != member_by_id.end()) {
+    member = pos->second;
+    return DDS::RETCODE_OK;
+  } else {
+    return DDS::RETCODE_ERROR;
+  }
+}
+
+void DynamicType::get_all_members(DynamicTypeMembersById& member)
+{
+  member = member_by_id;
+}
+
+ACE_CDR::ULong DynamicType::get_member_count()
+{
+  return member_by_name.size();
+}
+
+DDS::ReturnCode_t DynamicType::get_member_by_index(DynamicTypeMember_rch& member, ACE_CDR::ULong index)
+{
+  if (index < member_by_index.size()) {
+    member = member_by_index[index];
+    return DDS::RETCODE_OK;
+  } else {
+    return DDS::RETCODE_ERROR;
+  }
+}
+
+bool DynamicType::equals(const DynamicType& other)
+{
+  return test_equality(*this, other);
+}
+
+bool test_equality(const DynamicType& lhs, const DynamicType& rhs)
+{
+//7.5.2.8.4 Operation: equals
+//Two types shall be considered equal if and only if all of their respective properties, as identified
+//in Table 54 above, are equal.
+
+//Note: We are comparing the TypeDescriptor even though the spec seems to say not to
```

Comment:
Indent looks off.
Add more detail here, if it's deviating from the spec then we should have a reason for that

Reply:
DynamicTypeMember
> 7.5.2.6.3 Operation: equals
> Two members shall be considered equal if and only if they belong to the same type and all of
> their respective properties, as identified in Table 52 above, are equal.

DynamicType

> 7.5.2.8.4 Operation: equals
> Two types shall be considered equal if and only if all of their respective properties, as identified
> in Table 54 above, are equal.

This is the way the spec defines "equals" for DynamicTypes and DynamicTypeMembers.

The way this is defined will lead to infinite loops which never resolve anything.  If the equality of two DynamicTypes is dependant on the equality of two DynamicTypeMembers, and the DynamicTypeMembers' equality is dependant on the equality of the two DynamicTypes, we have a very simple unending loop.

Additionally, even if this didn't force an infinite loop, a simple thought experiment of two different DynamicTypeMembers belonging to the same DynamicType would be "equal" according to the spec's definition of "equals", because the "properties" section does not include descriptors.

I made the assumption that the only reasonable way to determining "equals" would be to check the descriptors in a recursive way until we reach the very base of each type, which is what I use for my test.

We could either not implement equals, or we can overwrite the spec defined equals, and make a spec issue for clarification.

Reply:
Without a deep dive into the spec, it seems like there are two issues here.

- One issue is that it seems like the spec is missing some properties in tables 52 and 54. Either that or they mean more than the table's "properties" when they say properties... possibly properties + results of certain operations.
- The other issue is that "the same type" in their definition can't mean equality through the equals method, as you point out. So what does it mean? Same name? Same descriptor? I don't have an answer here, since I'm not sure I have a full grasp of all the various "properties" to choose from.

Reply:
> The way this is defined will lead to infinite loops which never resolve anything. If the equality of two DynamicTypes is dependant on the equality of two DynamicTypeMembers, and the DynamicTypeMembers' equality is dependant on the equality of the two DynamicTypes, we have a very simple unending loop.

Ignoring the implementation details, I think the definition of "equals" for `DynamicTypeMember` in the spec makes sense. I interpret 2 `DynamicTypeMember` objects as equal if they refer to the same member of the same type. So it makes sense that checking the equality of 2 `DynamicTypeMember` objects involving checking that they represent members from the same type.

For implementation, maybe we can check that 2 `DynamicTypeMember` objects belong to the same type by keeping the `TypeIdentifier` of the containing type within each `DynamicTypeMember` object. If 2 such objects have the same containing type's `TypeIdentifier` then they belong to the same type. With that, we may solve the infinite loop issue.